### PR TITLE
Package

### PR DESCRIPTION
**Pull request by csadorf**
This pull request introduces a multitude of improvements and changes:

 - Conversion of the build source code to package structure.
 - Support for YAML files in addition to JSON files to define the app and categories registries, as well as app metadata.
 - Support for the `$ref` syntax to link to externally hosted metadata.
 - Support for explicit static files.
 - The build process can be controlled via a configuration file (`config.yaml`).
 - Introduction of revised schemas (v2) to support new metadata syntax.
 - Introduction of pre-commit checks.

### CODE REVIEW

**Comment by csadorf** [comment]
@yakutovicha Please do not merge directly after the review, because I want to revert https://github.com/aiidalab/aiidalab-registry/pull/59/commits/85637fc694d299e9c79e9cde5258046309b61b30 and maybe remove the helper scripts from https://github.com/aiidalab/aiidalab-registry/pull/59/commits/8a1365fde8512785a86d4ab382cf30040fb6cf63 before merge.

**Review by yakutovicha — COMMENTED** [review]
Thanks, @csadorf, awesome work. I have a couple of minor comments/questions.

1. Why did you put the old and the new schemas in different places? Maybe I don't see what you try to achieve but it feels to me that having different versions in one place would be better.
2. I think we should be consistent and call it App Store, not registry. Can be done in a different PR, though. The confusion is there due to historical reasons.
3. Found "AiiDA Lab" that should be renamed AiiDAlab. Again, can be done in a different PR.
4. In the new apps.yaml you explicitly define metadata, but in the readme, you talk about `$ref` to the external documents. Is there a preferred way? If yes, can you add one-two sentences to the readme about it? Or elaborate on which way is preferred under which conditions.

File: apps.yaml
Comment (csadorf) [reply apps.yaml:6]: Yes, we need to use explicit declaration until the metadata files are updated in the app repositories. The metadata schema is largely the same, but I removed the `version` key, because it does not make much sense. We should also wait with these updates until #60 is finalized.

**Comment by csadorf** [comment]
>     1. Why did you put the old and the new schemas in different places? Maybe I don't see what you try to achieve but it feels to me that having different versions in one place would be better.

The old schemas are no longer "used", but instead still put on the website for future reference. In principle we should not remove previously published schemas, because they might still be referenced by outdated AiiDAlab deployments etc. That's why they are placed as "static" files and not in the `schemas/` folder.

> 
>     2. I think we should be consistent and call it App Store, not registry. Can be done in a different PR, though. The confusion is there due to historical reasons.
> 

Yeah, this annoys me a lot, however it might make sense to distinguish between the registry (`apps.yaml`) and the actual `App store` where you can install stuff within AiiDAlab. In other words, the registry is really only a registry, the app store is a user-facing application.

>     3. Found "AiiDA Lab" that should be renamed AiiDAlab. Again, can be done in a different PR.

I'll just fix it in this PR.
 
>     4. In the new apps.yaml you explicitly define metadata, but in the readme, you talk about `$ref` to the external documents. Is there a preferred way? If yes, can you add one-two sentences to the readme about it? Or elaborate on which way is preferred under which conditions.

The reason is that we need to update the remote references first before we can use `$ref` as they would no longer be compliant. See also my other comment regarding this.

File: apps.yaml
Comment (csadorf) [reply apps.yaml:6]: I just realized that in this PR, I have actually not changed the metadata definition yet. So I am going use `$ref` instead.

File: src/mod/templates/base.html
Comment (yakutovicha) [inline src/mod/templates/base.html:29]:
```suggestion
            <a href="{{ aiidalab_url }}">AiiDAlab</a> application registry
```

File: src/mod/templates/base.html
Comment (yakutovicha) [inline src/mod/templates/base.html:41]:
```suggestion
    This page is part of the <a href="{{ app_registry_url }}">AiiDAlab registry</a>, the official registry of <a href="{{ aiidalab_url }}">AiiDAlab</a> applications.
```

File: src/mod/templates/main_index.html
Comment (yakutovicha) [inline src/mod/templates/main_index.html:54]:
```suggestion
                <!-- Summaryinfo has not yet been introduced to the AiiDAlab app registry -->
```

**Review by yakutovicha — COMMENTED** [review]
I think it should be called `AiiDAlab` rather than `AiiDALab`

**Comment by csadorf** [comment]
> I think it should be called `AiiDAlab` rather than `AiiDALab`

You're right. 🤦

File: migrate.py
Comment (csadorf) [reply migrate.py]: I've removed the migration-related scripts from this PR they are strictly not needed here and are merely a distraction. I'll re-introduce them again separately.

File: schemas/apps.schema.json
Comment (csadorf) [reply schemas/apps.schema.json:45]: I've addressed the rename, but suggest that the broader "app store" vs "app registry" change is handled in a separate PR.

**Comment by yakutovicha** [comment]
> The reason is that we need to update the remote references first before we can use `$ref` as they would no longer be compliant. See also my other comment regarding this.

Ok, this is clear. Essentially, before people update their `metadata.json` we provide a temporary replacement that works, right?

**Comment by yakutovicha** [comment]
> Yeah, this annoys me a lot, however it might make sense to distinguish between the registry (`apps.yaml`) and the actual `App store` where you can install stuff within AiiDAlab. In other words, the registry is really only a registry, the app store is a user-facing application.

To be honest, I am not sure this separation is really necessary. The part of the problem is that it is not fully clear where to put a bar when differentiating what is a registry and what is a store. Is it only `apps.json`, or `apps_meta.json` as well? How would we call https://aiidalab.github.io/aiidalab-registry/ then?

**Comment by csadorf** [comment]
> > The reason is that we need to update the remote references first before we can use `$ref` as they would no longer be compliant. See also my other comment regarding this.
> 
> Ok, this is clear. Essentially, before people update their `metadata.json` we provide a temporary replacement that works, right?

Exactly. But only if it's necessary of course.

**Comment by csadorf** [comment]
> > Yeah, this annoys me a lot, however it might make sense to distinguish between the registry (`apps.yaml`) and the actual `App store` where you can install stuff within AiiDAlab. In other words, the registry is really only a registry, the app store is a user-facing application.
> 
> To be honest, I am not sure this separation is really necessary. The part of the problem is that it is not fully clear where to put a bar when differentiating what is a registry and what is a store. Is it only `apps.json`, or `apps_meta.json` as well? How would we call https://aiidalab.github.io/aiidalab-registry/ then?

@yakutovicha I suggest we have a separate discussion about this, but try to merge this PR into develop now. It should be ready.

**Comment by yakutovicha** [comment]
> The old schemas are no longer "used", but instead still put on the website for future reference. In principle we should not remove previously published schemas, because they might still be referenced by outdated AiiDAlab deployments etc. That's why they are placed as "static" files and not in the `schemas/` folder.

Actually, my comment was not about keeping the old schemas. The need of that for backwards compatibility is quite clear. 

I was more referring to the new schemas. Why aren't they placed in `src/static/schemas/v2`? The new schema standards are also static in the sense that whenever one wants to modify them, one should better make a new schema version, I guess.

**Comment by csadorf** [comment]
> I was more referring to the new schemas. Why aren't they placed in `src/static/schemas/v2`? The new schema standards are also static in the sense that whenever one wants to modify them, one should better make a new schema version, I guess.

Agreed, I will try that.

**Review by yakutovicha — APPROVED** [review]
great work, thanks @csadorf

**Pull request merged**
